File: DeductionCpp/Abstraction/Domain.cpp
#include "Domain.h"
#include "Connectives/And.h"
#include "Connectives/Or.h"
#include "Connectives/Not.h"
#include "Connectives/Implication.h"
#include "Connectives/Equivalence.h"
#include "Connectives/And.h"
#include "PropositionSymbol.h"
#include "PropositionArray.h"

using namespace std;

namespace DeductionCpp { namespace Abstraction {
    Domain* Domain::DomainInstance = 0;

    Domain::Domain()
    {
        Domain::DomainInstance = this;

        this->Members = std::vector<DomainMember>();
        this->Members.push_back(DomainMember('&', DomainMemberTypes::BinaryConnective, 1, &Connectives::And::CreateInstance));
        this->Members.push_back(DomainMember('|', DomainMemberTypes::BinaryConnective, 2, &Connectives::Or::CreateInstance));
        this->Members.push_back(DomainMember('!', DomainMemberTypes::UnaryConnective, 0, &Connectives::Not::CreateInstance));
        this->Members.push_back(DomainMember('>', DomainMemberTypes::BinaryConnective, 3, &Connectives::Implication::CreateInstance));
        this->Members.push_back(DomainMember('=', DomainMemberTypes::BinaryConnective, 3, &Connectives::Equivalence::CreateInstance));

        this->Members.push_back(DomainMember('f', DomainMemberTypes::Constant, 0, &PropositionSymbol::CreateInstance));
        this->Members.push_back(DomainMember('t', DomainMemberTypes::Constant, 0, &PropositionSymbol::CreateInstance));

        this->Members.push_back(DomainMember('(', DomainMemberTypes::Array, 0, &PropositionArray::CreateInstance));
    }

    Domain& Domain::Instance() {
        if (Domain::DomainInstance == 0)
        {
            new Domain();
            atexit(&Domain::Dispose);
        }

        return *Domain::DomainInstance;
    }

    void Domain::Dispose()
    {
        delete Domain::DomainInstance;
        Domain::DomainInstance = 0;
    }

    DomainMember* Domain::GetMemberBySymbolChar(char symbolChar)
    {
        for (auto it = this->Members.begin(); it != this->Members.end(); it++)
        {
            if (it->SymbolChar == symbolChar)
            {
                return &(*it);
            }
        }

        return NULL;
    }
} }

File: Deduction/Abstraction/Domain.cs
using System;
using System.Collections.Generic;
using Deduction.Abstraction.DomainMembers;

namespace Deduction.Abstraction
{
    public class Domain
    {
        protected static Domain instance = null;
        protected readonly List<DomainMember> members;

        public List<DomainMember> Members
        {
            get
            {
                return this.members;
            }
        }

        public static Domain Instance
        {
            get
            {
                if (Domain.instance == null)
                {
                    Domain.instance = new Domain();
                }

                return Domain.instance;
            }
        }

        protected Domain()
        {
            this.members = new List<DomainMember>()
            {
                new DomainMember("&", typeof(And), 2, null),
                new DomainMember("|", typeof(Or), 3, null),
                new DomainMember("!", typeof(Not), 1, null),
                new DomainMember(">", typeof(Implication), 4, null),
                new DomainMember("=", typeof(Equivalence), 4, null),

                new DomainMember("f", typeof(PropositionSymbol), 0, false),
                new DomainMember("t", typeof(PropositionSymbol), 0, true)
            };
        }

        public static DomainMember GetMemberByType(Type type)
        {
            foreach (DomainMember domainMember in Domain.Instance.Members)
            {
                if (domainMember.Type == type)
                {
                    return domainMember;
                }
            }

            return null;
        }

        public static DomainMember GetMemberBySymbolChar(string symbolChar)
        {
            foreach (DomainMember domainMember in Domain.Instance.Members)
            {
                if (domainMember.SymbolChar == symbolChar)
                {
                    return domainMember;
                }
            }

            return null;
        }

        public static DomainMember[] GetMembersSorted()
        {
            DomainMember[] domainMembers = Domain.Instance.Members.ToArray();

            Array.Sort(
                domainMembers,
                (DomainMember x, DomainMember y) =>
                {
                    if (x.Precedence > y.Precedence)
                    {
                        return 1;
                    }

                    if (y.Precedence > x.Precedence)
                    {
                        return -1;
                    }

                    return 0;
                }
            );

            return domainMembers;
        }
    }
}
